youtube_transcript_api/_api.py

```diff
@@ -99,47 +99,45 @@ def get_transcript(cls, video_id, languages=None, proxies=None):
 
 class _TranscriptFetcher():
     WATCH_URL = 'https://www.youtube.com/watch?v={video_id}'
-    API_BASE_URL = 'https://www.youtube.com/api/{api_url}'
-    LANGUAGE_REGEX = re.compile(r'(&lang=.*&)|(&lang=.*)')
+    API_BASE_URL = 'https://www.youtube.com/api/'
     TIMEDTEXT_STRING = 'timedtext?v='
+    NAME_REGEX = re.compile(r'(&name=.*&)|(&name=.*)')
 
     def __init__(self, video_id, languages, proxies):
         self.video_id = video_id
         self.languages = languages
+        print(languages)
         self.proxies = proxies
 
     def fetch(self):
         if self.proxies:
             fetched_site = requests.get(self.WATCH_URL.format(video_id=self.video_id), proxies=self.proxies).text
         else:
             fetched_site = requests.get(self.WATCH_URL.format(video_id=self.video_id)).text
-        timedtext_splits = fetched_site.split(self.TIMEDTEXT_STRING)
-        timedtext_url_start = (
-            timedtext_splits[2].find(self.TIMEDTEXT_STRING)
-            + len(timedtext_splits[0])
-            + len(timedtext_splits[1])
-            + len(self.TIMEDTEXT_STRING) + 1
-        )
-
-        for language in (self.languages if self.languages else [None,]):
-            response = self._execute_api_request(fetched_site, timedtext_url_start, language)
+        timedtext_splits = [split[:split.find('"')]
+                .replace('\\u0026', '&')
+                .replace('\\', '') 
+                for split in fetched_site.split(self.TIMEDTEXT_STRING)]
+        matched_splits = []
+        for language in self.languages:
+            matched_splits = [split for split in timedtext_splits if '&lang={}'.format(language) in split]
+            if matched_splits:
+                break
+        if matched_splits:
+            timedtext_url = min(matched_splits, key=self._sort_splits)
+            response = self._execute_api_request(timedtext_url)
             if response:
                 return response
 
         return None
 
-    def _execute_api_request(self, fetched_site, timedtext_url_start, language):
-        url = self.API_BASE_URL.format(
-            api_url=fetched_site[
-                timedtext_url_start:timedtext_url_start + fetched_site[timedtext_url_start:].find('"')
-            ].replace(
-                '\\u0026', '&'
-            ).replace(
-                '\\', ''
-            )
-        )
-        if language:
-            url = re.sub(self.LANGUAGE_REGEX, '&lang={language}&'.format(language=language), url)
+    #Sorting the matched splits by string length because we want non-asr options returned first
+    #However, we don't want to include the length of the 'name' argument as it could possible throw this off
+    def _sort_splits(self, matched_split):
+        return len(re.sub(self.NAME_REGEX, r'\1', matched_split))
```

Comment:
Wouldn't it be most reliable to just check if the string contains `kind=asr` and then multiply the length by 1 and 0 otherwise? Or are there other `matched_split`s which don't contain a proper URL at all?

Reply:
the advantage of this system is that it will return variations or dialects of a desired language. For example, if `en` is desired `en-GB` will still return but only if an `en` track is not available

Reply:
But if the name for the transcript is longer than `kind=asr`, it would be sorted below it. Not sure if that is actually possible tho 🤔.
You could do a combination of both just to be extra sure: 
```python
def _sort_splits(self, matched_split):
        return len(re.sub(self.NAME_REGEX, r'\1', matched_split)) ** (2 if 'kind=asr' in matched_split else 1)
```

Reply:
@danielcliu what do you think about this?

Reply:
@jdepoix Sorry for the very late reply.

> But if the name for the transcript is longer than kind=asr, it would be sorted below it.

The `NAME_REGEX` should be removing the 'name' url parameter from the url, so this case should never happen.

Reply:
@danielcliu But `lang` can also have a different length than 2. For example danielcliu uzin2GZPJY0 has a transcript with `lang=es-419`. In the case of this video it's no problem, but the value could potentially be longer. Also the param `signature` does not have a fixed size. Therefore it feels to me, that size is not metric which is reliable 100% of the time. Although it probably is reliable more than 99% of the time....

I think I might have found a way more reliable way, but I'll have to put a bit more time into it. I will try to implement this in the next couple of days and push it into your PR.

Reply:
It doesn't really matter if the `lang` is longer than 2 chars as long as it isn't greater than 9 (the len('&kind=asr')). I guess this is possible but haven't seen any `language` tags that long. Looking through the documentation that I believe youtube is [following](https://tools.ietf.org/html/bcp47#appendix-B) it would likely be a `language-variant` or the like to be long enough to break the algorithm, which isn't something I have found youtube to use.

>   Also the param signature does not have a fixed size. 

I haven't checked this from video to video, but I haven't seen a video's caption urls have different signature length in them. Do you have an example?

>Either way, thank you very much your contribution and your endurance in this PR!

Yeah I was hoping to get this resolved much quicker, but I have been busy with other things/messed up and forgot to respond to some of your comments for a while. So thanks for the patience

Reply:
I agree that it is fairly unlikely, but it seems possible in theory. Doesn't matter however, as I have found a way more elegant solution to this! 😄👍 
I will push it into your PR and then you can let me know what you think of it. It will also enable easier implementation of future features like retrieving a list of available subtitles and automatically translating subtitles into other languages!